### PR TITLE
Spanish translation

### LINKED ISSUE #511: Translation of labels to other languages
Hi!

I've been spending a lot of time trying to translate to spanish some of the predetermined labels or footers that appears on tables made with gtsummary.

I've used some workaorunds more or less like:

```r
nice_table <- some_data %>%
 tbl_summary()

nice_table[[2]] <- nice_table[[2]] %>% 
  mutate(footnote=str_replace(footnote, pattern = "Statistics presented:", replacement = 
 "Estadísticos:"),  
  footnote=str_replace(footnote, pattern = "median", replacement = "mediana:"))

nice_table %>% 
  as_gt() %>% 
  tab_options(table.font.size = 12) %>%
  cols_label("label"=md("**Característica**") #Here I'm translating the "Characteristics" header
```
So I was guessing that if you are interested, gtsummary could have an option to offer those labels in spanish. 

I don't have the programming level (or knowledge) to make this by a pull request but if you give me a list with all of the possible labels or footnotes that gtsummary displays then I'll give you the translations to spanish so you can modify the code to add the translations.

Finally the user could select the language to be displayed passing an argument in a way such tbl_summary(lan="spanish|español").

I'm sure that this could be a great feature for your package and that the +400 millions of spanish speakers will surely appreciate it! ;)

Thanks for your attention and best regards!

### PR DESCRIPTION
**Pull request by oranwutang**
**What changes are proposed in this pull request?**

Created a csv with statistical terms translated to spanish, also I added a file with instructions on how to handle the csv in case the user is not used to work with files that contain UTF-8 characters

closes #511

### CODE REVIEW

**Comment by ddsjoberg** [comment]
HI @oranwutang !

I think we've made some pretty quick progress here! Wonderful! I added our to do list from our email here.  For the time-being, let's focus only on updating `tbl_summary()` and its related functions.  Once we've ironed out the details for `tbl_summary()`, we can continue implementing for the rest of the package.

- [ ] Create pull request from orangutan/gtsummary@spanish_translation into ddsjoberg/gtsummary@master TODO: @oranwutang 
- [ ] Convert our excel file of translations into a CSV file, and add it to \gtsummary\data-raw\ folder.  TODO: @oranwutang
- [ ] Continue to fill out the translation data sheet by adding the translation of all statistical tests performed with add_p().  TODO: @oranwutang
- [ ] Create a language theme TODO: Daniel
- [ ] Create a new function that accepts two args: language, and English text.  If the language is English, the function returns the passed text. If another language, the function goes to our translation lookup table, finds the translations and returns it.  If there is no translation for the passed text, the function will return the English text. TODO: Daniel
- [ ] Execute the function above throughout the internals of the package: TODO: Daniel and @oranwutang
- [ ] Benchmark speed after implementing the language theme.  I am concerned the current implementation may slow down the function.  We should run some tests of speed for the current code and compare to the code in the master branch, and then to the version on CRAN.  @oranwutang do you have any experience with this? I've never done it before.

**Comment by oranwutang** [comment]
Wow, the result is perfect! You're right this is progressing very quickly.

I'll take care of the benchmark and post the results as soon as posible.

**Comment by oranwutang** [comment]
Hi @ddsjoberg ,

I'm trying to install the spanish_translation version via:

`remotes::install_github("oranwutang/gtsummary@spanish_translation", force = TRUE)`

Downloading GitHub repo oranwutang/gtsummary@spanish_translation
✓  checking for file ‘/tmp/RtmpKb2ChF/remotes16129bf67ae/oranwutang-gtsummary-15f4b21/DESCRIPTION’ ...
─  preparing ‘gtsummary’:
✓  checking DESCRIPTION meta-information ...
─  installing the package to process help pages
─  checking for LF line-endings in source and make files and shell scripts (4.9s)
─  checking for empty or unneeded directories
─  building ‘gtsummary_1.3.0.9019.tar.gz’
   
Installing package into ‘/home/rstudio-user/R/x86_64-pc-linux-gnu-library/4.0’
(as ‘lib’ is unspecified)
* installing *source* package ‘gtsummary’ ...
** using staged installation
** R
** data
*** moving datasets to lazyload DB
** inst
** byte-compile and prepare package for lazy loading
** help
*** installing help indices
*** copying figures
** building package indices
** installing vignettes
** testing if installed package can be loaded from temporary location
** testing if installed package can be loaded from final location
** testing if installed package keeps a record of temporary installation path
* DONE (gtsummary)

But it throws the following errors and warnings:

`library(gtsummary)`

Error: package or namespace load failed for ‘gtsummary’ in get(method, envir = home):
 lazy-load database '/home/rstudio-user/R/x86_64-pc-linux-gnu-library/4.0/gtsummary/R/gtsummary.rdb' is corrupt
In addition: Warning messages:
1: In .registerS3method(fin[i, 1], fin[i, 2], fin[i, 3], fin[i, 4],  :
  restarting interrupted promise evaluation
2: In get(method, envir = home) :
  restarting interrupted promise evaluation
3: In get(method, envir = home) : internal error -3 in R_decompress1

Do you know how to solve it?

**Comment by oranwutang** [comment]
Forget it! by now it is working

**Comment by oranwutang** [comment]
Hi!

I've made the benchmark

Here are the results:

table

Rplot


What do you think?

**Comment by ddsjoberg** [comment]
closes #511

File: R/theme_gtsummary.R [hunk 1]
```diff
@@ -122,3 +122,16 @@ theme_gtsummary_printer <- function(
 
   list("pkgwide-str:print_engine" = match.arg(print_engine))
 }
+
+# ------------------------------------------------------------------------------
+#' @rdname theme_gtsummary
+#' @param language String indicating language. Must be one of `"es"` (Spanish)
+#' or `"en"` (English).
+#' @export
+theme_gtsummary_language <- function(language = c("es", "en")) {
```
Comment (karissawhiting) [inline R/theme_gtsummary.R:131]: we discussed that you could potentially remove `en`, but not necessary.

File: R/utils-tbl_summary.R [hunk 2]
```diff
@@ -655,26 +655,27 @@ tbl_summary_input_checks <- function(data, by, label, type, value, statistic,
 # stat_label_match -------------------------------------------------------------
 # provide a vector of stat_display and get labels back i.e. {mean} ({sd}) gives Mean (SD)
 stat_label_match <- function(stat_display, iqr = TRUE) {
+  language <- get_theme_element("pkgwide-str:language", default = "en")
```
[the remark this answers is not in the document]
Comment (oranwutang) [reply R/utils-tbl_summary.R:658]: I think that line could be deleted without problems

File: R/tbl_regression.R [hunk 3]
```diff
@@ -333,17 +333,18 @@ estimate_header <- function(x, exponentiate) {
   }
 
   # assigning header and footer ------------------------------------------------
+  language <- get_theme_element("pkgwide-str:language", default = "en")
```
[the remark this answers is not in the document]
Comment (oranwutang) [reply R/tbl_regression.R:336]: I agree, language arguments should be deleted

File: R/theme_gtsummary.R (hunk 1 above)
Comment (oranwutang) [reply R/theme_gtsummary.R:131]: I think that the `en` option should be keeped, is the only way (at least I did not find another) to keep the choice for the user to get back to english if the language has been changed

**Comment by oranwutang** [comment]
Hi @karissawhiting , I'm eager to see gtsummary speaking italian!

File: R/theme_gtsummary.R (hunk 1 above)
Comment (ddsjoberg) [reply R/theme_gtsummary.R:131]: Ahh, that is a good point!

**Comment by ddsjoberg** [comment]
My colleague added French to the translations CSV file.
- The accents appear in the CSV file
- She saved the CSV with UTF-8 encoding
- The accents accurately render in the gtsummary tables
- I ran the checks and it passed the non-ASCII characters check!
- I will submit the package to CRAN's internal check system to double check we won't run into as issue with them when we submit.
- @oranwutang if the CRAN checks pass, perhaps we can update the Spanish to include the accented characters to make the file easier on the eyes?

image

File: R/tbl_regression.R [hunk 4]
```diff
@@ -333,17 +333,18 @@ estimate_header <- function(x, exponentiate) {
...
   if (model_type == "logistic") {
     header <- ifelse(exponentiate == TRUE, "OR", "log(OR)")
-    attr(header, "footnote") <- "OR = Odds Ratio"
```
Comment (ddsjoberg) [inline R/tbl_regression.R:338]: @oranwutang @karissawhiting what do you think about allowing the abbreviations to be updated as well?  I think that should be ok, right?

File: R/tbl_summary.R [hunk 5]
```diff
@@ -175,8 +175,9 @@ tbl_summary <- function(data, by = NULL, label = NULL, statistic = NULL,
   value <- value %||% get_theme_element("tbl_summary-arg:value")
   missing <- missing %||% get_theme_element("tbl_summary-arg:missing",
                                             default = "ifany")
-  missing_text <- missing_text %||% get_theme_element("tbl_summary-arg:missing_text",
-                                                      default = "Unknown")
+  missing_text <- missing_text %||%
+    get_theme_element("tbl_summary-arg:missing_text",
```
Comment (ddsjoberg) [inline R/tbl_summary.R:179]: @oranwutang i added translation for the "Unknown" in `tbl_summary()` output.  Can you add the Spanish translation?

**Comment by ddsjoberg** [comment]
FYI with the updated translation CSV, all the CRAN checks passed

File: R/tbl_regression.R (hunk 4 above)
Comment (oranwutang) [reply R/tbl_regression.R:338]: At least in spanish it won't be necessary, as a matter of facts, in Spanish OR is usuallly abbreviated "OR" and although the translation is "razón de probabilidades" many people use "odds ratio" as well (same with hazard ratio)

File: R/tbl_summary.R (hunk 5 above)
Comment (oranwutang) [reply R/tbl_summary.R:179]: Done!

**Comment by oranwutang** [comment]
Hi!

I have corrected the accents on the csv file, maybe they were damaged while adding the french translations, I did forget 👏 to tell you by e-mail that depending on the software used to edit encoded files you must indicate the encoding while opening the file (not just while saving it), bad mine! 

I have tested the new version and everything seems to work just fine! 🍻

image

image

**Comment by ddsjoberg** [comment]
Oops @oranwutang , you absolutely told me not to open in excel 🤦 🤦 🤦 sorry about that!

**Pull request merged**